### PR TITLE
Add Shelley keypair generation

### PR DESCRIPTION
**Pull request by intricate**

### CODE REVIEW

File: cardano-api/src/Cardano/Api/Crypto/Shelley.hs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import           Crypto.Random (MonadRandom)
+
+import           Shelley.Spec.Ledger.Keys (SKey (..), pattern VKey)
+
+genKeyPair :: MonadRandom m
+           => m KeyPair
+genKeyPair = do
+  sk <- genKeyDSIGN
+  pure $ KeyPairShelley (VKey (deriveVerKeyDSIGN sk)) (SKey sk)
```
Comment (intricate) [inline cardano-api/src/Cardano/Api/Crypto/Shelley.hs:22]: Perhaps this should be in `cardano-ledger-specs` instead.

File: cardano-api/src/Cardano/Api/Types.hs [hunk 2]
```diff
@@ -35,8 +41,8 @@ data KeyPair
   -- An 'XPub' is 32 bytes of public key and 32 bytes of Chaincode which is used in the
   -- Byron address derivation scheme.
   = KeyPairByron !Crypto.VerificationKey !Crypto.SigningKey
-  | KeyPairShelley -- Exact crypto not ready yet.
-  deriving (Generic, NFData, Show)
+  | KeyPairShelley !(VKey TPraosStandardCrypto) !(SKey TPraosStandardCrypto)
```
Comment (intricate) [inline cardano-api/src/Cardano/Api/Types.hs:44]: Should I use the `cardano-ledger-specs` types (`VKey` and `SKey`) or stick with `VerKeyDSIGN (DSIGN TPraosStandardCrypto)` and `SignKeyDSIGN (DSIGN TPraosStandardCrypto)`?

File: cardano-api/src/Cardano/Api/Crypto/Shelley.hs (hunk 1 above)
Comment (erikd) [reply cardano-api/src/Cardano/Api/Crypto/Shelley.hs:22]: Is that a separate file? If so I think its so small and trivial i should just be in `Cardano.Api`.

File: cardano-api/src/Cardano/Api.hs [hunk 3]
```diff
@@ -78,7 +82,7 @@ mkPublicKey :: KeyPair -> Network -> PublicKey
 mkPublicKey kp nw =
   case kp of
     KeyPairByron vk _ -> PubKeyByron nw vk
-    KeyPairShelley -> PubKeyShelley
+    KeyPairShelley _ _ -> PubKeyShelley
```
Comment (erikd) [inline cardano-api/src/Cardano/Api.hs:85]: `PubKeyShelley` needs two fields.

File: cardano-api/src/Cardano/Api/CBOR.hs [hunk 4]
```diff
@@ -54,15 +60,23 @@ keyPairFromCBOR bs =
       tag <- CBOR.decodeWord8
       case tag of
         172  -> KeyPairByron <$> fromCBOR <*> fromCBOR
-        173  -> pure KeyPairShelley
+        173  -> do
+          decodedVerKey <- decodeVerKeyDSIGN
+          decodedSignKey <- decodeSignKeyDSIGN
+          pure $ KeyPairShelley (VKey decodedVerKey) (SKey decodedSignKey)
```
Comment (erikd) [inline cardano-api/src/Cardano/Api/CBOR.hs:66]: No love for `Applicative`.

File: cardano-api/src/Cardano/Api/CBOR.hs [hunk 5]
```diff
@@ -6,6 +8,11 @@ module Cardano.Api.CBOR
   , publicKeyFromCBOR
   , publicKeyToCBOR
 
+  , decodeShelleyVerificationKey
```
Comment (erikd) [inline cardano-api/src/Cardano/Api/CBOR.hs:11]: I find `'decode` and `encode` a little too imprecise. That is why I named these as `typenameToCBOR` and `typenameFromCBOR` .

File: cardano-api/src/Cardano/Api/CBOR.hs [hunk 6]
```diff
@@ -6,6 +8,11 @@ module Cardano.Api.CBOR
...
+  , encodeShelleyVerificationKey
```
Comment (erikd) [inline cardano-api/src/Cardano/Api/CBOR.hs:12]: Probably do not need to export the `encode `and `decode` functions.

Comment (intricate) [reply cardano-api/src/Cardano/Api/CBOR.hs:12]: Sure, I'll drop them from the export list.

File: cardano-api/test/Test/Cardano/Api/Gen.hs [hunk 7]
```diff
@@ -38,16 +51,61 @@ genAddress =
 
 genKeyPair :: Gen KeyPair
 genKeyPair =
-  -- When Shelly is sorted out, this should change to `Gen.choose`.
-  Gen.frequency
-    [ (9, genKeyPairByron)
-    , (1, pure KeyPairShelley)
+  Gen.choice
```
Comment (erikd) [inline cardano-api/test/Test/Cardano/Api/Gen.hs:54]: :+1:

File: cardano-api/src/Cardano/Api.hs [hunk 8]
```diff
@@ -10,6 +11,8 @@ module Cardano.Api
   , KeyPair (..)
   , Network (..)
   , PublicKey (..)
+  , ShelleyKeyDiscriminator (..)
+  , ShelleyVerificationKey (..)
```
Comment (erikd) [inline cardano-api/src/Cardano/Api.hs:15]: Its a pity that these leak from the API, but lets get the API finished and then see if we can clean it up.

**Pull request merged**